scripts/midi_remap.lua

```diff
@@ -0,0 +1,103 @@
+ardour {
+    ["type"]    = "dsp",
+    name        = "MIDI Remap Notes",
+    category    = "Utility",
+    license     = "MIT",
+    author      = "Alby Musaelian",
+    description = [[Map arbitrary MIDI notes to others. Affects Note On/Off and polyphonic key pressure.]]
+}
+
+-- The number of remapping pairs to allow. Increasing this (at least in theory)
+-- decreases performace, so it's set fairly low as a default. The user can
+-- increase this if they have a need to.
+N_REMAPINGS = 10
+
+function dsp_ioconfig ()
+    return { { midi_in = 1, midi_out = 1, audio_in = 0, audio_out = 0}, }
+end
+
+function dsp_params ()
+    local note_names = {"C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"}
+
+    local map_scalepoints = {}
+    for note=0,127 do
+        local octave = -5 + math.floor(note / 12)
+        local name = note_names[(note % 12) + 1]
+        map_scalepoints[string.format("%03d (%s%d)", note, name, octave)] = note
+    end
+
+    local map_params = {}
+
+    i = 1
+    for mapnum = 1,N_REMAPINGS do
+        -- From and to
+        for _,name in pairs({"| #" .. mapnum .. "  Map note", "|__   to"}) do
+            map_params[i] = {
+                ["type"] = "input",
+                name = name,
+                min = 0,
+                max = 127,
+                default = 0,
+                integer = true,
+                enum = true,
+                scalepoints = map_scalepoints
+            }
+            i = i + 1
+        end
+    end
+
+    return map_params
+end
+
+function dsp_init(samplerate)
+    -- Init translation table, reserve memory
+    translation_table = {}
+    for i = 0,127 do
+        translation_table[i] = i
+    end
+end
+
+
+function dsp_run (_, _, n_samples)
+    assert (type(midiin) == "table")
+    assert (type(midiout) == "table")
+    local cnt = 1;
+
+    function tx_midi (time, data)
+        midiout[cnt] = {}
+        midiout[cnt]["time"] = time;
+        midiout[cnt]["data"] = data;
+        cnt = cnt + 1;
+    end
+
+    -- We build the translation table every buffer because, as far as I can tell,
+    -- there's no way to only rebuild it when the parameters have changed.
+    -- As a result, it has to be updated every buffer for the parameters to have
+    -- any effect.
+
+    -- Restore translation table
+    for i = 0,127 do
+        translation_table[i] = i
+    end
+    -- Update it
+    local ctrl = CtrlPorts:array()
+    for i=1,N_REMAPINGS*2,2 do
+        translation_table[ctrl[i]] = ctrl[i + 1]
```

Comment:
If there are duplicate mappings, the last one wins.

This is mainly relevant for the default  `000 -> n`.  Try re-mapping note 0 using any but the last control.
One solution would be to add a dedicated  `"off" == -1` case as control parameter.

Reply:
Since I'm switching to the sparse arrays, I'll add an dedicated "off" option